Domains: Remove UsersStore in favor of `react-query`

tyxla opened the pull request:
#### Changes proposed in this Pull Request

With #51784 landing, now we have a single use of the users store in the `TransferOtherUser` component - part of the flow to transfer a domain to another user. This PR removes that last instance and does some cleanup. Full list of the changes:

* Introduce a simple `withUser` higher-order component to allow us to use `react-query` without having to refactor a component to a functional one.
* Refactor `TransferOtherUser` to use `withUser` instead of the legacy users store.
* Cleanup `DomainManagementData` from any legacy users store handling, as it's no longer used anywhere.
* Cleanup from `StoreConnection` and remove it completely.

This PR is part of #24150 and the unified reduxification (de-fluxification) effort. 😉 

#### Testing instructions

* Verify `/domains/manage/:domain/transfer/other-user/:site` still works like it did before.
* Specifically, verify it correctly loads the users that are eligible for a domain transfer.

#### Notes

I'm pinging some teams for review:
* @Automattic/cobalt and @aidvu - because this touches Domains
* @Automattic/team-calypso-frameworks - because this PR removes more Flux!

tyxla committed:
Data: Introduce a withUsers HoC

tyxla committed:
Domains: Refactor TransferOtherUser to use react-query users

tyxla committed:
Domains: Get rid of needsUsers and UsersStore

tyxla committed:
Store: Remove flux store listeners from StoreConnection

tyxla committed:
Components: README fixes

flootr commented on client/data/users/with-users.js:
```diff
@@ -0,0 +1,25 @@
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+const withUsers = createHigherOrderComponent(
+	( Wrapped, fetchOptions = {}, queryOptions = {} ) => ( props ) => {
```
Not sure but do we ever need to make `fetchOptions` and `queryOptions` available to the outside?

tyxla replied:
Perhaps, yes. Imagine if we wanted to search by a term or provide a limit, then those would be useful. But because we're not using those yet, I'm happy to remove them if you'd prefer.

flootr replied:
That makes sense although I think constructing those objects should be part of the HOC or in general the component which use the query hook. We can iterate later though, no need to remove those now.

flootr commented on client/data/users/with-users.js:
```diff
@@ -0,0 +1,25 @@
+const withUsers = createHigherOrderComponent(
+	( Wrapped, fetchOptions = {}, queryOptions = {} ) => ( props ) => {
+		const siteId = useSelector( getSelectedSiteId );
+		const usersQuery = useUsersQuery( siteId, fetchOptions, queryOptions );
```
`useUsersQuery` is using `useInfiniteQuery` under the hood which provides pagination handlers. Do we ever need to call [`fetchNextPage`](https://react-query.tanstack.com/reference/useInfiniteQuery#_top)? I'm not exactly sure about previous behaviour.

tyxla replied:
Not for this use case, but we might need it in the future.

flootr commented on client/data/users/with-users.js:
```diff
@@ -0,0 +1,25 @@
+ * External dependencies
+ */
+import React from 'react';
+import { createHigherOrderComponent } from '@wordpress/compose';
```
TIL 👍🏻

sirbrillig commented:
I have no problems with this, although I'm curious about `because of the cleanups in StoreConnection`. Is there something specific we're using `StoreConnection` for that I should be testing?

flootr commented:
Testing this I found a regression which might be related to the number of users we're fetching. For some sites, I don't see all the users in the list which I see on prod.

Previously we seemed to have fetched 1000 users (see [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/components/data/domain-management/index.jsx#L80)) and now we're only fetching 100 ([default](https://github.com/Automattic/wp-calypso/blob/trunk/client/data/users/use-users-query.js#L13) for `useUsersQuery`). Those users are then [filtered](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/domains/domain-management/transfer/transfer-to-other-user/index.jsx#L336) for users with the role `administrator` which might cause this inconsistency. Changing this to fetch 1000 users might fix this but is there a way to only fetch users we're interested in by telling the API in some way?

jsnajdr commented:
> is there a way to only fetch users we're interested in by telling the API in some way?

The `/sites/:site/users` endpoint support a `role` query arg that you can use to do the `{ role: 'administrator' }` filter on the server.

I found this by looking at the endpoint's PHP source. It should also be visible in the API console, but I didn't find a way how to see it there 😕

flootr commented:
The Rest API documentation for WP.com has some information on that as well: https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/users/

tyxla committed:
Move HoC inline, add role query arg

tyxla commented:
> I have no problems with this, although I'm curious about `because of the cleanups in StoreConnection`. Is there something specific we're using `StoreConnection` for that I should be testing?

I think I was mis-blaming something, plus GH suggested that I ping you specifically for review (likely because you touched those files before). Feel free to ignore this ping 😉

tyxla committed:
Remove unnecessary filtering by admin

tyxla commented:
@flootr @jsnajdr as you suggested, we're now filtering at the the API request. I've also moved the HoC inline and refactored it a little bit - the extra args I was previously providing to `createHigherOrderComponent` didn't work as I was expecting.

Please, feel free to take another look when you get a chance.

jsnajdr commented on client/components/data/domain-management/index.jsx:
```diff
@@ -117,7 +88,6 @@ class DomainManagementData extends React.Component {
 						selectedDomainName={ this.props.selectedDomainName }
 						selectedSite={ selectedSite }
 						sitePlans={ this.props.sitePlans }
-						stores={ stores }
```
It seems to me that the `StoreConnection` component lost all meaning by now and can be removed:
- doesn't maintain any store subscriptions
- the `isDataLoading` and `loadingPlaceholder` props are not used
- `getStateFromStores` merely passes props, with no local state left

Simply rendering `this.props.component` with the same props as `StoreConnection` had will do the same job. `domains` prop is passed conditionally, and `currentUser` prop is renamed to `user`, and all other props are simply passed through 1:1.

tyxla replied:
Agreed! I was planning to clean that up in another PR, but don't mind doing it here. Cleaned up in the last 2 commits - https://github.com/Automattic/wp-calypso/pull/51809/commits/5ab49ae45c6edbf2c28e3048bab260a1d4295c7a and  https://github.com/Automattic/wp-calypso/pull/51809/commits/dfc1325ed9a5a74e168eece5271a87684dcb0a3d

tyxla committed:
Bypass StoreConnection

tyxla committed:
Remove unused StoreConnection

jsnajdr commented on client/my-sites/domains/domain-management/transfer/transfer-to-other-user/index.jsx:
```diff
@@ -334,18 +338,25 @@ class TransferOtherUser extends React.Component {
-				includes( user.roles, 'administrator' ) &&
-				this.getWpcomUserId( user ) !== this.props.currentUser.ID
-		);
+		return users.filter( ( user ) => this.getWpcomUserId( user ) !== this.props.currentUser.ID );
```
I think there is a pre-existing bug here that's worth fixing. We should exclude also site users who don't have a wpcom account (`linked_user_ID === false`) because we can't transfer a domain to someone who's not a wpcom user.

The UI that lets you transfer the site ownership to someone else, very similar to domain transfer in principle, has code that excludes these users correctly:

https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/site-settings/manage-connection/site-ownership.jsx#L46-L53

See also my another observation about checking `linked_user_ID`: https://github.com/Automattic/wp-calypso/commit/2c332d34027d97ec1da28bbe564b4c618fd2ec95#r49403087

tyxla replied:
Good catch. Fixed in https://github.com/Automattic/wp-calypso/pull/51809/commits/e360ace6cc119d684a1b9af75029587c9416923e

jsnajdr commented on client/my-sites/domains/domain-management/transfer/transfer-to-other-user/index.jsx:
```diff
@@ -334,18 +338,25 @@ class TransferOtherUser extends React.Component {
 
+const withUsers = createHigherOrderComponent(
+	( Wrapped ) => ( props ) => {
+		const siteId = useSelector( getSelectedSiteId );
```
The `TransferToOtherUser` component already receives a `selectedSite` prop from above (from `DomainManagementData`) and uses it heavily. The `withUsers` wrapper should use the same source of truth and look at `props.selectedSite.ID` instead of doing its own Redux selection.

tyxla replied:
I believe keeping this as-is makes the component more flexible, and retrieving the selected site ID is cheap on its own. Also, I don't see much value in passing the ID down in this case. 

Do you feel strongly about this?

jsnajdr replied:
I don't feel strongly about it and it's not a big issue. But I'd personally do the change 🙂 The `withUsers` wrapper should eventually be merged into the `TransferToOtherUser` component when it becomes functional.

tyxla replied:
Well, I don't intend to convert it at this time, who knows, it may remain a `Component` for a while 😉  I guess this is one of those things that we can do eventually, but at this time they don't matter so much.

tyxla committed:
Exclude users without a WP.com account

jsnajdr commented on client/components/data/domain-management/index.jsx:
```diff
@@ -105,20 +61,17 @@ class DomainManagementData extends React.Component {
+						isRequestingSiteDomains: this.props.isRequestingSiteDomains,
+						products: this.props.products,
+						selectedDomainName: this.props.selectedDomainName,
+						selectedSite: this.props.selectedSite,
```
nit: `selectedSite` is already destructured a few lines above.

tyxla replied:
> nit: `selectedSite` is already destructured a few lines above.

Addressed in #51856.

jsnajdr reviewed (approved):
There are a few unwanted things that could be easily eliminated from the `TransferToOtherUser` component:
- the `get` usage in `getWpcomUserId`. The method could be shortened to mere `user.linked_user_ID ?? user.ID`, completely eliminating the `isAtomic` selector and prop
- the `UNSAFE_componentWillUpdate` method that sets `selectedUserId` to the first user on the list if it doesn't have a better value. The logic could be moved either to the `setState` call (there's only one) or to the code that reads the state value.
- improving the `selectedUserId` logic could also remove the `_.head` usage

Approving because I no longer have any blockers to address, just optional ideas for more improvement.

flootr reviewed (approved):
This now tests well for me 👍🏻 I leave it to you which of the other suggestions you want to implement now or follow up with.

tyxla commented:
Thank you for the great feedback, friends 🙌 ! 

I'm going to address those in a follow-up PR, to keep this one more contained.

tyxla commented:
> There are a few unwanted things that could be easily eliminated from the `TransferToOtherUser` component:
> 
> * the `get` usage in `getWpcomUserId`. The method could be shortened to mere `user.linked_user_ID ?? user.ID`, completely eliminating the `isAtomic` selector and prop
> * the `UNSAFE_componentWillUpdate` method that sets `selectedUserId` to the first user on the list if it doesn't have a better value. The logic could be moved either to the `setState` call (there's only one) or to the code that reads the state value.
> * improving the `selectedUserId` logic could also remove the `_.head` usage

Thanks for those suggestions, I've addressed them in #51855.
